src/encoded/static/components/matrix_chip_seq.js

```diff
@@ -0,0 +1,886 @@
+import { svgIcon } from '../libs/svg-icons';
+import * as globals from './globals';
+import { MatrixInternalTags } from './objectutils';
+import { SearchFilter } from './matrix';
+import { TextFilter, SearchControls } from './search';
+import DataTable from './datatable';
+
+
+const VISUALIZE_LIMIT = 500;
+const MIN_SIZE_OF_LARGE_MATRIX = 250;
+const SEARCH_PERFORMED_PUBSUB = 'searchPerformed';
+const SEARCH_RESULT_COUNT_PUBSUB = 'searchResultCount';
+const CLEAR_SEARCH_BOX_PUBSUB = 'clearSearchBox';
+
+/**
+ * Tranform context to a form where easier to fetch information
+ *
+ * @param {context} context - Context from react
+ * @param {string} assayTitle - Assay Title
+ * @param {string} organismName - Organism Name
+ * @returns {object} - Object with structure - { chIPSeqData, subTabs };
+ *
+ *      chIPSeqData is an object where:
+ *          key is a sub Tab
+ *          value is of structure - { headerRow, dataRow, assayTitle, organismName };
+ *              headerRow - header content-list
+ *              dataRow - array of non-header content
+ *                  key: biosample ontology classification
+ *                  value: array containing counts
+ *              assayTitle - Assay title
+ *              organismName- Organism name
+ *      subTabs: List of subTabs for easy access
+ */
+const getChIPSeqData = (context, assayTitle, organismName) => {
+    if (!context || !context.matrix || !context.matrix.x || !context.matrix.y || !assayTitle || !organismName) {
+        return null;
+    }
+
+    const subTabSource = 'biosample_ontology.classification';
+    const subTabs = context.matrix.x[subTabSource].buckets.map(x => x.key);
+    const chIPSeqData = {};
+
+    subTabs.forEach((subTab) => {
+        const xGroupBy1 = context.matrix.x.group_by[0];
+        const xGroupBy2 = context.matrix.x.group_by[1];
+        const headerRow = context.matrix.x[xGroupBy1].buckets.find(f => f.key === subTab)[xGroupBy2]
+            .buckets
+            .reduce((a, b) => {
+                const m = a.concat(b);
+                return m;
+            }, [])
+            .map(x => x.key);
+        const headerRowIndex = headerRow.reduce((x, y, z) => { x[y] = z; return x; }, []);
+        const headerRowLength = headerRow.length;
+        const yGroupBy1 = context.matrix.y.group_by[0];
+        const yGroupBy2 = context.matrix.y.group_by[1];
+
+        const yData = context.matrix.y[yGroupBy1].buckets
+            .find(rBucket => rBucket.key === organismName)[yGroupBy2].buckets
+            .reduce((a, b) => {
+                const m = {};
+                m[b.key] = b[xGroupBy1].buckets
+                    .filter(f => f.key === subTab)
+                    .reduce((x, y) => {
+                        x.push(y[xGroupBy2].buckets
+                            .reduce((i, j) => i.concat(j), []));
+                        return x;
+                    }, []);
+                return a.concat(m);
+            }, []);
```

Comment:
This looks like a lot of work; I wonder if it would make sense to push this to the backend?

Reply:
I can think about changing the format of data from elastic search. For things like subtabs-creation , searching and filtering out empty rows, to do these quickly and frequently on tabs, I need the data in a format that backend did not provide. That is why I had to dig into it like this.

Reply:
From an architectural perspective we want to keep things that change a lot like frontend views isolated from things that change less like core search logic. Since this data transformation is very tied to a frontend view, and specifically an encoded frontend view, it would make sense to keep it isolated in a data transformation layer between the search API and React rendering.

The question is then should this layer be on the Python or the Javascript side. I doubt there would be any performance benefits to keeping it on the Python side and the amount of data transferred would be more or less the same as now. Also since this code is doing transformations to render tabs and rows it seems more appropriate to keep it on the JS side rather than piping it through a JSON API in Python.

So why not just make a data transformation layer in JS with reusable library code?

Reply:
well this page specifically has performance issues with large matrices (no idea if this loop has any bearing on that).

Reply:
@hitz reading and processing of the data from the backend is pretty quick. It is the painting/DOM update that is the slow part.